net/ipv4/tcp_output.c

```diff
@@ -2494,6 +2571,14 @@ static int tcp_mtu_probe(struct sock *sk)
 	}
 	tcp_init_tso_segs(nskb, nskb->len);
 
+#ifdef CONFIG_SECURITY_TEMPESTA
+	result = tcp_tfw_sk_write_xmit(sk, nskb, mss_now);
```

Comment:
Here, you silently break PMTU discovery logic. `tcp_mtu_probe` must prepare and send only one segment, the size of the segment must be equal to `probe_size`, but `probe_size` will be greater than `mss_now`. 

Introducing the call to `tcp_tfw_sk_write_xmit()` you break this rules. In `tcp_tfw_sk_write_xmit()` you using cong or send window to set the limit for `tfw_tls_encrypt()` instead of  `probe_size`, then in `tfw_tls_encrypt()` you are trying to merge another skbs into one while not reaching the limit. And in the end of `tcp_tfw_sk_write_xmit()` you are reducing effective size of the segment doing `tcp_set_skb_tso_segs(skb, mss_now)`.

Reply:
Fixed